### PR TITLE
feat: add segments

### PR DESCRIPTION
**Pull request by olav**

### CODE REVIEW

File: src/lib/types/model.ts [hunk 1]
```diff
@@ -341,3 +341,12 @@ export interface IProjectWithCount extends IProject {
     featureCount: number;
     memberCount: number;
 }
+
+export interface ISegment {
+    id: number;
+    name: string;
+    description: string;
```
Comment (sighphyre) [inline src/lib/types/model.ts:348]: The table def seems to allow nulls on this field, should this be nullable?

File: src/lib/types/model.ts [hunk 2]
```diff
@@ -341,3 +341,12 @@ export interface IProjectWithCount extends IProject {
...
+    constraints: IConstraint[];
+    createdBy: string;
```
Comment (sighphyre) [inline src/lib/types/model.ts:350]: Same for this field

File: src/test/e2e/api/admin/segment.e2e.test.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+    const [segmentB] = await fetchSegments();
+    expect(segmentB.id).toEqual(segmentA.id);
+    expect(segmentB.name).toEqual('b');
+    expect(segmentB.createdAt).toBeDefined();
+    expect(segmentB.constraints.length).toEqual(0);
+});
+
+test('should update segment constraints', async () => {
+    const constraintA = { contextName: 'a', operator: 'IN' };
```
Comment (sighphyre) [inline src/test/e2e/api/admin/segment.e2e.test.ts:150]: I might just be reflecting my lack of knowledge of how we want this to work but this test suggests that we can make constraints with no value or values (I think this is wrong!) I think that's being caused by existing code not validating that the a constraint has a value xor values. Just something to think about in the coming weeks - if we patch that it'll break these tests

Comment (olav) [reply src/test/e2e/api/admin/segment.e2e.test.ts]: Good point. I'll add `values` to the test data at least.

**Review by chriswk — CHANGES_REQUESTED** [review]
Some concerns about the migrations, but I like the api and functions added :+1:

File: perf/test/gzip.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+set -feu
+
+cd "$(dirname "$0")"
+
+source ../env.sh
+
+print_response_size () {
+    local URL="$1"
+    local RES="$(curl -s -H "Authorization: $PERF_AUTH_KEY" "$URL")"
```
[the remark this answers is not in the document]
Comment (olav) [reply perf/test/gzip.sh]: @sonatype-lift ignore

File: perf/test/artillery.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
+
+set -feu
+
+cd "$(dirname "$0")"
+
+. ../env.sh
```
[the remark this answers is not in the document]
Comment (olav) [reply perf/test/artillery.sh:7]: @sonatype-lift ignore

File: perf/test/gzip.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+set -feu
+
+cd "$(dirname "$0")"
+
+source ../env.sh
```
[the remark this answers is not in the document]
Comment (olav) [reply perf/test/gzip.sh]: @sonatype-lift ignore

File: src/lib/routes/admin-api/constraints.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        }: Pick<IUnleashServices, 'featureToggleServiceV2'>,
+    ) {
+        super(config);
+        this.featureService = featureToggleServiceV2;
+        this.logger = config.getLogger('/admin-api/validation.ts');
+
+        this.post('/validate', this.validateConstraint, NONE);
+    }
+
+    async validateConstraint(
```
Comment (FredrikOseberg) [inline src/lib/routes/admin-api/constraints.ts:28]: Wasn't this previously implemented in another controller? If it's changed, have we updated the FE counterpart?

File: src/lib/services/segment-schema.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import joi from 'joi';
+import { constraintSchema } from '../schema/feature-schema';
+
+export const segmentSchema = joi
+    .object()
+    .keys({
+        name: joi.string().required(),
+        description: joi.string().allow(null).allow('').optional(),
```
Comment (FredrikOseberg) [inline src/lib/services/segment-schema.ts:8]: Could this be an issue in more strictly typed languages? If something can be a null or a string?

File: src/lib/routes/admin-api/constraints.ts (hunk 7 above)
Comment (olav) [reply src/lib/routes/admin-api/constraints.ts:28]: Right. We needed this endpoint for the segments page, where we don't have a projectId or featureName for the path. 

The frontend path is updated in the segments branch.

File: src/lib/services/segment-schema.ts (hunk 8 above)
Comment (olav) [reply src/lib/services/segment-schema.ts:8]: Looks like we usually do `.allow(null).allow('').optional()` for optional string fields.

Comment (FredrikOseberg) [reply src/lib/services/segment-schema.ts:8]: True. And it makes sense if we never expose this in the client API. But if we for some reason want to expose fields like this in the client API in the future we might have a problem. Is there any reason we can't default it to empty string instead?

**Review by chriswk — APPROVED** [review]
I'm excited :)

File: perf/env.sh [hunk 9]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+export PERF_AUTH_KEY="*:*.964a287e1b728cb5f4f3e0120df92cb5"
+export PERF_APP_URL="http://localhost:4242"
```
[the remark this answers is not in the document]
Comment (olav) [reply perf/env.sh:4]: Nice catch! Could we use the auto-created `PERF_AUTH_KEY` for artillery as well? One less key to manage.

File: src/lib/types/partial.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+// Recursively mark all properties as optional.
+export type PartialDeep<T> = T extends object
+    ? {
+          [P in keyof T]?: PartialDeep<T[P]>;
+      }
+    : T;
+
+// Mark one or more properties as optional.
+export type PartialSome<T, K extends keyof T> = Pick<Partial<T>, K> &
```
[the remark this answers is not in the document]
Comment (olav) [reply src/lib/types/partial.ts:9]: Same here, at least the API is simple enough. Might be built into TS at some point.

**Pull request merged**